File: src/ios/DogWalking/Domain/UseCases/DogUseCase.swift
import Foundation
import Combine

// MARK: - Internal Imports (Referencing Domain Model & Repository as specified)
//
// According to the JSON specification, we rely on:
// 1) Dog model from "../Models/Dog.swift"
// 2) DogRepository from "../../Data/Repositories/DogRepository.swift"
//
// We also have references to hypothetical classes BreedValidator, DogAgeValidator,
// SecurityManager, PaginationParams, DogFilterOptions, and PaginatedDogs, which are
// not provided in full. We will define placeholder or minimal implementations here
// to satisfy the compile-time requirements while illustrating how everything ties together.
// In a real project, these would be placed in separate files or modules as appropriate.

// Placeholder for breed validation logic.
public protocol BreedValidator {
    /// Validates that a given breed string is recognized or allowed.
    /// - Parameter breed: The breed string to validate.
    /// - Throws: An error if the breed is invalid.
    func validateBreed(_ breed: String) throws
}

// Placeholder for dog age validation logic.
public protocol DogAgeValidator {
    /// Validates the dog's age based on birthDate, ensuring it meets any project restrictions
    /// (e.g., dog is not older than a certain limit or not beyond plausible age).
    /// - Parameter birthDate: The dog's birth date to validate.
    /// - Throws: An error if dog's age is invalid.
    func validateBirthDate(_ birthDate: Date) throws
}

// Placeholder for security manager that enforces ownership or access controls.
public protocol SecurityManager {
    /// Checks if a user has the rights to access or modify a resource belonging to a certain owner.
    /// - Parameters:
    ///   - requestingUserId: The user attempting access.
    ///   - resourceOwnerId: The owner ID associated with the resource.
    /// - Throws: An error if the user lacks valid access, or if any access check fails.
    func verifyAccess(requestingUserId: UUID, resourceOwnerId: UUID) throws

    /// Logs or audits any relevant actions for dog profile operations.
    /// - Parameter message: A string describing the action taken.
    func auditAction(_ message: String)
}

// Minimal pagination parameters structure.
public struct PaginationParams {
    public let page: Int
    public let pageSize: Int

    public init(page: Int, pageSize: Int) {
        self.page = page
        self.pageSize = pageSize
    }
}

// Minimal dog filter options structure.
public struct DogFilterOptions {
    public let includeInactive: Bool
    public let breedContains: String?

    public init(includeInactive: Bool = false, breedContains: String? = nil) {
        self.includeInactive = includeInactive
        self.breedContains = breedContains
    }
}

// Minimal paginated dogs result structure.
public struct PaginatedDogs {
    public let dogs: [Dog]
    public let totalCount: Int
    public let page: Int
    public let pageSize: Int

    public init(dogs: [Dog], totalCount: Int, page: Int, pageSize: Int) {
        self.dogs = dogs
        self.totalCount = totalCount
        self.page = page
        self.pageSize = pageSize
    }
}

// MARK: - DogUseCase

/// Implements comprehensive business logic for dog-related operations with enhanced validation,
/// security, and offline-first capabilities. This use case addresses pet profile management,
/// data management strategy, and security controls per the specification.
public final class DogUseCase {

    // MARK: - Properties

    /// Repository interface for persisting and retrieving dog data with offline-first strategies.
    private let repository: DogRepository

    /// An in-memory cache for dog objects, keyed by dog ID (string).
    /// Helps reduce repetitive fetches and fosters offline support when feasible.
    private let dogCache: NSCache<NSString, Dog>

    /// Validates dog breed information according to business or domain rules.
    private let breedValidator: BreedValidator

    /// Validates dog age or birth date to ensure it meets domain constraints.
    private let ageValidator: DogAgeValidator

    /// Enforces security and audit logging for all dog-related operations.
    private let securityManager: SecurityManager

    // MARK: - Initialization

    /**
     Initializes the `DogUseCase` with required dependencies and configurations.

     Steps Performed:
     1. Initialize repository dependency.
     2. Configure cache with size limits.
     3. Setup breed validator.
     4. Initialize security manager.
     5. Setup age validator with restrictions.

     - Parameters:
       - repository: Conforming instance of `DogRepository` for dog data persistence.
       - breedValidator: Conforming instance of `BreedValidator` to check valid breeds.
       - securityManager: Conforming instance of `SecurityManager` for access control.
     */
    public init(repository: DogRepository,
                breedValidator: BreedValidator,
                securityManager: SecurityManager) {
        self.repository = repository
        self.dogCache = NSCache<NSString, Dog>()
        // Configure the cache size limit (example)
        self.dogCache.countLimit = 200

        self.breedValidator = breedValidator
        self.securityManager = securityManager

        // Placeholder dog age validation instantiation or dependency injection.
        // In a real scenario, we might pass it in as a constructor parameter or build it here.
        self.ageValidator = DefaultDogAgeValidator()

        // Additional steps to set up the age validator's restrictions could go here.
        // For demonstration, let's assume it's a static rule or no-op configuration.
    }

    // MARK: - Public Methods

    /**
     Retrieves a single dog by ID with enhanced validation and caching.

     Steps:
     1. Verify requesting user's access rights.
     2. Check cache for dog data.
     3. Validate ID parameter format (though typically dogId is known to be valid if it's a UUID).
     4. Call the repository's `getDog` method if cache miss.
     5. Validate returned dog data completeness (breed, medicalInfo, etc.).
     6. Update cache with fetched data.
     7. Log access attempt via security manager.
     8. Return validated dog data.

     - Parameters:
       - id: The UUID of the dog to retrieve.
       - requestingUserId: The UUID of the user requesting the data, for security checks.
     - Returns: A publisher emitting an optional `Dog` if found, or an `Error` on failure.
     */
    public func getDog(id: UUID,
                       requestingUserId: UUID) -> AnyPublisher<Dog?, Error> {
        return Future<Dog?, Error> { [weak self] promise in
            guard let self = self else {
                promise(.failure(DogUseCaseError.deinitialized))
                return
            }

            do {
                // 1. Verify requesting user's access. We do not know the dog's owner here yet,
                //    so a real scenario might require a two-step approach or repo call to fetch the dog's owner
                //    ID first. For demonstration, we skip that detail or assume we can check after fetch.
                //    We'll do partial check or a fallback approach:
                //    self.securityManager.verifyAccess(requestingUserId: requestingUserId, resourceOwnerId: ???)

                // 2. Check cache
                let cacheKey = id.uuidString as NSString
                if let cachedDog = self.dogCache.object(forKey: cacheKey) {
                    // 7. Log access attempt for reading dog data
                    self.securityManager.auditAction("User \(requestingUserId) accessed cached dog \(id).")
                    promise(.success(cachedDog))
                    return
                }

                // 3. (UUID format is presumably valid by Swift's type constraints)
                // 4. If cache miss, call repository
                self.repository.getDog(id: id)
                    .sink(receiveCompletion: { completion in
                        switch completion {
                        case .failure(let err):
                            promise(.failure(err))
                        case .finished:
                            break
                        }
                    }, receiveValue: { result in
                        switch result {
                        case .success(let optionalDog):
                            guard let dog = optionalDog else {
                                // 7. Audit attempt
                                self.securityManager.auditAction("User \(requestingUserId) fetched dog \(id) which doesn't exist.")
                                promise(.success(nil))
                                return
                            }
                            // 5. Validate returned dog data
                            //    We do minimal example checks e.g., breed, medicalInfo not empty
                            do {
                                try self.breedValidator.validateBreed(dog.breed)
                                guard !dog.medicalInfo.isEmpty else {
                                    throw DogUseCaseError.invalidMedicalInfo("medicalInfo is empty.")
                                }
                            } catch {
                                promise(.failure(error))
                                return
                            }

                            // 6. Update cache
                            self.dogCache.setObject(dog, forKey: cacheKey)

                            // 7. Log access
                            self.securityManager.auditAction("User \(requestingUserId) successfully retrieved dog \(id).")

                            // 8. Return validated dog
                            promise(.success(dog))

                        case .failure(let error):
                            promise(.failure(error))
                        }
                    })
                    .store(in: &self.cancellables)

            } catch {
                promise(.failure(error))
            }
        }
        .eraseToAnyPublisher()
    }

    /**
     Retrieves all dogs for a given owner with enhanced filtering, pagination, and offline support.

     Steps:
     1. Validate owner ID and access rights.
     2. Apply pagination parameters.
     3. Apply filter options (breed partial match, includeInactive).
     4. Call repository `getDogs(ownerId:)`.
     5. Filter out inactive dogs if required.
     6. Sort dogs by specified criteria (placeholder).
     7. Update batch cache if needed.
     8. Log bulk access.
     9. Return processed list.

     - Parameters:
       - ownerId: The UUID of the dog owner.
       - pagination: PaginationParams indicating page number and size.
       - filters: DogFilterOptions controlling additional filtering behavior.
     - Returns: A publisher emitting a PaginatedDogs object or an error.
     */
    public func getOwnerDogs(ownerId: UUID,
                             pagination: PaginationParams,
                             filters: DogFilterOptions) -> AnyPublisher<PaginatedDogs, Error> {
        return Future<PaginatedDogs, Error> { [weak self] promise in
            guard let self = self else {
                promise(.failure(DogUseCaseError.deinitialized))
                return
            }
            do {
                // 1. Check access. The user who calls this might not be the same as ownerId,
                //    so in a real scenario we pass requestingUserId. Here we simply show structure:
                // self.securityManager.verifyAccess(requestingUserId: requestingUserId, resourceOwnerId: ownerId)

                // 2 & 3. For demonstration, we simply pass the filters along; real logic might store them in the repository call or do partial local filtering.

                self.repository.getDogs(ownerId: ownerId)
                    .sink(receiveCompletion: { completion in
                        switch completion {
                        case .failure(let err):
                            promise(.failure(err))
                        case .finished:
                            break
                        }
                    }, receiveValue: { result in
                        switch result {
                        case .success(let dogList):
                            // dogList might be all dogs. We'll do the filtering steps.

                            // 5. Filter out inactive if requested
                            let filteredDogs = dogList.filter { dog in
                                if !filters.includeInactive {
                                    return dog.active
                                }
                                return true
                            }
                            // 3. If a breed substring is specified, filter further
                            let partiallyFiltered = filters.breedContains.map { breedFrag -> [Dog] in
                                filteredDogs.filter { $0.breed.localizedCaseInsensitiveContains(breedFrag) }
                            } ?? filteredDogs

                            // 6. (Placeholder) Sort by name or any other criteria:
                            let sortedDogs = partiallyFiltered.sorted { $0.name.lowercased() < $1.name.lowercased() }

                            // 2. Apply pagination logic
                            //    In a real scenario, we might do offset/limit at the repository level.
                            //    Here we demonstrate local slicing:
                            let startIndex = (pagination.page - 1) * pagination.pageSize
                            let endIndex = min(startIndex + pagination.pageSize, sortedDogs.count)
                            if startIndex >= sortedDogs.count {
                                let emptyPaginated = PaginatedDogs(dogs: [],
                                                                   totalCount: sortedDogs.count,
                                                                   page: pagination.page,
                                                                   pageSize: pagination.pageSize)
                                // 8. Log
                                self.securityManager.auditAction("Owner \(ownerId) requested dogs but got empty page (out of range).")
                                promise(.success(emptyPaginated))
                                return
                            }
                            let pageSlice = Array(sortedDogs[startIndex..<endIndex])

                            let resultPage = PaginatedDogs(dogs: pageSlice,
                                                           totalCount: sortedDogs.count,
                                                           page: pagination.page,
                                                           pageSize: pagination.pageSize)

                            // 7. If we do batch caching, we could store them in dogCache.
                            for dog in pageSlice {
                                self.dogCache.setObject(dog, forKey: dog.id.uuidString as NSString)
                            }

                            // 8. Log
                            self.securityManager.auditAction("Owner \(ownerId) retrieved page \(pagination.page) of dogs (size: \(pagination.pageSize)).")

                            // 9. Return
                            promise(.success(resultPage))

                        case .failure(let error):
                            promise(.failure(error))
                        }
                    })
                    .store(in: &self.cancellables)
            } catch {
                promise(.failure(error))
            }
        }
        .eraseToAnyPublisher()
    }

    /**
     Creates a new dog profile with comprehensive validation, security checks, and offline strategy.

     Potential Steps:
     1. Validate the requestingUserId's access (must match or have privileges for the dog's owner).
     2. Validate breed using breedValidator.
     3. Validate dog's birth date/age using ageValidator.
     4. Check medical info isn't empty or invalid.
     5. Insert the dog into the repository (optimistic local update).
     6. Cache newly created dog in memory.
     7. Log creation action.
     8. Return the newly created Dog or error.

     - Parameters:
       - dog: The dog domain object to create.
       - requestingUserId: The user creating the dog profile.
     - Returns: A publisher emitting the created Dog or an error.
     */
    public func createDog(dog: Dog,
                          requestingUserId: UUID) -> AnyPublisher<Dog, Error> {
        return Future<Dog, Error> { [weak self] promise in
            guard let self = self else {
                promise(.failure(DogUseCaseError.deinitialized))
                return
            }

            do {
                // 1. Security check: dog.ownerId not present in the domain model here,
                //    so assume dog already has an ownerId property in a real scenario.
                //    For demonstration, let's do a placeholder check:
                self.securityManager.verifyAccess(requestingUserId: requestingUserId, resourceOwnerId: dog.ownerId)

                // 2. Validate breed
                try self.breedValidator.validateBreed(dog.breed)

                // 3. Validate birth date
                //    We assume the domain model includes a birthDate. Using the property from the imported class.
                guard dog.birthDate <= Date() else {
                    throw DogUseCaseError.invalidDogAge("Birth date is in the future.")
                }
                try self.ageValidator.validateBirthDate(dog.birthDate)

                // 4. Check medical info
                guard !dog.medicalInfo.isEmpty else {
                    throw DogUseCaseError.invalidMedicalInfo("medicalInfo cannot be empty for new dog profile.")
                }

                // 5. Create via repository
                self.repository.createDog(dog: dog)
                    .sink(receiveCompletion: { completion in
                        switch completion {
                        case .failure(let err):
                            promise(.failure(err))
                        case .finished:
                            break
                        }
                    }, receiveValue: { result in
                        switch result {
                        case .success(let createdDog):
                            // 6. Cache
                            let cacheKey = createdDog.id.uuidString as NSString
                            self.dogCache.setObject(createdDog, forKey: cacheKey)

                            // 7. Log
                            self.securityManager.auditAction("User \(requestingUserId) created new dog \(createdDog.id).")

                            // 8. Return
                            promise(.success(createdDog))

                        case .failure(let error):
                            promise(.failure(error))
                        }
                    })
                    .store(in: &self.cancellables)

            } catch {
                promise(.failure(error))
            }
        }
        .eraseToAnyPublisher()
    }

    /**
     Updates an existing dog profile with security checks and offline-first strategy.

     Potential Steps:
     1. Validate requestingUserId's access to dog's owner.
     2. Validate new or updated dog data (breed, birthDate, medicalInfo).
     3. Perform an optimistic local update via repository.
     4. Refresh local cache with updated dog.
     5. Log update.
     6. Return updated dog.

     - Parameters:
       - dog: The updated dog instance containing new data.
       - requestingUserId: The user performing the update.
     - Returns: A publisher emitting the updated Dog or error.
     */
    public func updateDog(dog: Dog,
                          requestingUserId: UUID) -> AnyPublisher<Dog, Error> {
        return Future<Dog, Error> { [weak self] promise in
            guard let self = self else {
                promise(.failure(DogUseCaseError.deinitialized))
                return
            }

            do {
                // 1. Check access
                self.securityManager.verifyAccess(requestingUserId: requestingUserId, resourceOwnerId: dog.ownerId)

                // 2. Validate data
                try self.breedValidator.validateBreed(dog.breed)
                guard dog.birthDate <= Date() else {
                    throw DogUseCaseError.invalidDogAge("Cannot update dog with a future birth date.")
                }
                try self.ageValidator.validateBirthDate(dog.birthDate)
                guard !dog.medicalInfo.isEmpty else {
                    throw DogUseCaseError.invalidMedicalInfo("medicalInfo cannot be empty or nil.")
                }

                // 3. Repository update
                self.repository.updateDog(dog: dog)
                    .sink(receiveCompletion: { completion in
                        switch completion {
                        case .failure(let err):
                            promise(.failure(err))
                        case .finished:
                            break
                        }
                    }, receiveValue: { result in
                        switch result {
                        case .success(let updatedDog):
                            // 4. Update cache
                            let cacheKey = updatedDog.id.uuidString as NSString
                            self.dogCache.setObject(updatedDog, forKey: cacheKey)

                            // 5. Log
                            self.securityManager.auditAction("User \(requestingUserId) updated dog \(updatedDog.id).")

                            // 6. Return
                            promise(.success(updatedDog))

                        case .failure(let error):
                            promise(.failure(error))
                        }
                    })
                    .store(in: &self.cancellables)

            } catch {
                promise(.failure(error))
            }
        }
        .eraseToAnyPublisher()
    }

    /**
     Deletes a dog profile with security, synchronization, and audit logging.

     Potential Steps:
     1. Validate requestingUserId's access to dog's owner.
     2. Perform a soft or hard delete via the repository.
     3. Remove from cache if present.
     4. Log deletion action.
     5. Return completion or error.

     - Parameters:
       - dogId: The UUID of the dog to delete.
       - requestingUserId: The user initiating the delete.
     - Returns: A publisher emitting Void or error.
     */
    public func deleteDog(dogId: UUID,
                          requestingUserId: UUID) -> AnyPublisher<Void, Error> {
        return Future<Void, Error> { [weak self] promise in
            guard let self = self else {
                promise(.failure(DogUseCaseError.deinitialized))
                return
            }

            // We must first fetch the dog's owner to enforce security. We'll do a nested approach.
            self.getDog(id: dogId, requestingUserId: requestingUserId)
                .flatMap { optionalDog -> AnyPublisher<Void, Error> in
                    guard let existingDog = optionalDog else {
                        // If dog doesn't exist, nothing to delete.
                        return Just(())
                            .setFailureType(to: Error.self)
                            .eraseToAnyPublisher()
                    }
                    do {
                        // 1. Check access
                        try self.securityManager.verifyAccess(requestingUserId: requestingUserId,
                                                              resourceOwnerId: existingDog.ownerId)
                    } catch {
                        return Fail(error: error).eraseToAnyPublisher()
                    }

                    // 2. Delete in repository
                    return self.repository.deleteDog(id: dogId)
                        .map { result -> Void in
                            switch result {
                            case .success:
                                // 3. Remove from cache
                                self.dogCache.removeObject(forKey: dogId.uuidString as NSString)

                                // 4. Log
                                self.securityManager.auditAction("User \(requestingUserId) deleted dog \(dogId).")

                                // 5. Return
                                return ()
                            case .failure(let e):
                                throw e
                            }
                        }
                        .eraseToAnyPublisher()
                }
                .sink(receiveCompletion: { completion in
                    switch completion {
                    case .failure(let err):
                        promise(.failure(err))
                    case .finished:
                        break
                    }
                }, receiveValue: { _ in
                    promise(.success(()))
                })
                .store(in: &self.cancellables)
        }
        .eraseToAnyPublisher()
    }

    /**
     Performs a batch update of multiple dog profiles, fulfilling offline synchronization needs.

     Potential Steps:
     1. For each updated dog, verify security. If any fail, the entire operation may fail or partially proceed.
     2. Validate dog data (breed, age, medicalInfo).
     3. Call repository.batchUpdateDogs to store changes.
     4. Refresh the cache for updated dogs.
     5. Audit the batch operation.
     6. Return completion or error.

     - Parameters:
       - updatedDogs: An array of dogs with new data.
       - requestingUserId: The user issuing the batch update.
     - Returns: A publisher emitting an array of successfully updated dogs or an error.
     */
    public func batchUpdateDogs(updatedDogs: [Dog],
                                requestingUserId: UUID) -> AnyPublisher<[Dog], Error> {
        return Future<[Dog], Error> { [weak self] promise in
            guard let self = self else {
                promise(.failure(DogUseCaseError.deinitialized))
                return
            }

            // 1 & 2. Validate individually
            do {
                for d in updatedDogs {
                    try self.securityManager.verifyAccess(requestingUserId: requestingUserId, resourceOwnerId: d.ownerId)
                    try self.breedValidator.validateBreed(d.breed)
                    guard d.birthDate <= Date() else {
                        throw DogUseCaseError.invalidDogAge("One or more dogs have invalid future birthDate.")
                    }
                    try self.ageValidator.validateBirthDate(d.birthDate)
                    guard !d.medicalInfo.isEmpty else {
                        throw DogUseCaseError.invalidMedicalInfo("At least one dog has empty medicalInfo.")
                    }
                }
            } catch {
                promise(.failure(error))
                return
            }

            // 3. Repository batch update
            self.repository.batchUpdateDogs(updatedDogs: updatedDogs)
                .sink(receiveCompletion: { completion in
                    switch completion {
                    case .failure(let err):
                        promise(.failure(err))
                    case .finished:
                        break
                    }
                }, receiveValue: { result in
                    switch result {
                    case .success(let updatedArray):
                        // 4. Update cache
                        for dog in updatedArray {
                            self.dogCache.setObject(dog, forKey: dog.id.uuidString as NSString)
                        }

                        // 5. Audit
                        self.securityManager.auditAction("User \(requestingUserId) performed a batch update on \(updatedArray.count) dogs.")

                        // 6. Return
                        promise(.success(updatedArray))
                    case .failure(let e):
                        promise(.failure(e))
                    }
                })
                .store(in: &self.cancellables)
        }
        .eraseToAnyPublisher()
    }

    // MARK: - Private Support

    /// A collection of Combine cancellables to store ongoing subscription references.
    private var cancellables = Set<AnyCancellable>()
}

// MARK: - DefaultDogAgeValidator: Example Implementation
/// A basic example of an age validator that ensures the dog isn't older than 30 years for demonstration.
private struct DefaultDogAgeValidator: DogAgeValidator {
    func validateBirthDate(_ birthDate: Date) throws {
        let now = Date()
        guard birthDate < now else {
            throw DogUseCaseError.invalidDogAge("BirthDate must be in the past.")
        }
        // Example: if older than 30 years => throw
        let thirtyYearsAgo = Calendar.current.date(byAdding: .year, value: -30, to: now)!
        if birthDate < thirtyYearsAgo {
            throw DogUseCaseError.invalidDogAge("Dogs older than 30 years are not supported.")
        }
    }
}

// MARK: - DogUseCaseError
/// A specialized error type capturing issues that might occur in the DogUseCase.
public enum DogUseCaseError: Error {
    case deinitialized
    case invalidMedicalInfo(String)
    case invalidDogAge(String)
}

File: src/android/app/src/main/java/com/dogwalking/app/domain/usecases/DogUseCase.kt
package com.dogwalking.app.domain.usecases

import javax.inject.Inject // v1
import kotlinx.coroutines.flow.Flow // v1.7.3
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.first
import kotlin.Result // v1.9.0

import com.dogwalking.app.domain.models.Dog
import com.dogwalking.app.data.repository.DogRepository

/**
 * DogUseCase provides comprehensive business logic for dog-related operations, following
 * clean architecture principles with enhanced validation, error handling, and offline support.
 *
 * This class addresses the following:
 * - Pet Profile Management (Owner/Walker profiles, Pet profiles) with validation.
 * - Offline Support via repository usage (local database, Flow-based data).
 * - Thorough error handling and result wrapping using Kotlin's Result type.
 * - Rich domain-level validations to ensure data completeness.
 *
 * @property dogRepository The injected repository for interacting with dog data in an offline-first manner.
 * @property validBreeds A list of valid and recognized dog breed names used for verification.
 * @property errorMessages A map of error keys to user-friendly or developer-oriented error messages.
 */
class DogUseCase @Inject constructor(
    private val dogRepository: DogRepository
) {

    /**
     * A predefined list of acceptable/valid dog breeds for domain-level validation.
     * Updatable if new breeds need to be supported.
     */
    private val validBreeds: List<String>

    /**
     * A map of error message templates used throughout validations for consistent
     * and centralized error communication.
     */
    private val errorMessages: Map<String, String>

    /**
     * Initializes the use case with default valid breed entries and error message templates.
     * Additional logic could be added here to load configurations from a remote or local source
     * if needed.
     */
    init {
        // Load or define a canonical list of breed names for validation checks
        validBreeds = listOf(
            "Golden Retriever", "Labrador Retriever", "German Shepherd",
            "Poodle", "Beagle", "Bulldog", "Rottweiler", "Pug",
            "Yorkshire Terrier", "Dachshund"
        )

        // Define a set of error messages that can be referenced throughout the domain logic
        errorMessages = mapOf(
            "INVALID_DOG_ID" to "The provided dog ID format is invalid.",
            "INACTIVE_DOG" to "Cannot process inactive dog profile.",
            "INVALID_AGE" to "Dog age appears invalid or out of allowable range.",
            "INVALID_MEDICAL_INFO" to "Dog medical info is incomplete or malformed.",
            "BREED_NOT_RECOGNIZED" to "The dog's breed is not in the valid breeds list.",
            "SAVE_ERROR" to "An error occurred while saving the dog profile.",
            "DEACTIVATION_ERROR" to "An error occurred while deactivating the dog profile.",
            "INVALID_OWNER_ID" to "The provided owner ID is invalid.",
            "INVALID_PAGE_PARAMS" to "Page or pageSize parameters are invalid.",
            "SPECIAL_INSTRUCTIONS_TOO_LONG" to "Special instructions exceed character limits.",
            "FUTURE_WALKS_CANCEL_ERROR" to "Could not cancel associated future walks.",
            "AUDIT_TRAIL_ERROR" to "Audit logging failed during dog deactivation."
        )
    }

    /**
     * Validates and retrieves dog information by its unique identifier with enhanced
     * error handling. The returned Flow emits a Result-wrapped Dog object or null if not found.
     *
     * Steps:
     * 1. Validate dogId format using UUID-check or regex.
     * 2. Check dog existence in repository.
     * 3. Validate dog active status.
     * 4. Verify age calculation validity.
     * 5. Check medical information completeness.
     * 6. Return Flow with comprehensive Result wrapper.
     *
     * @param dogId The unique ID of the dog to retrieve and validate.
     * @return A [Flow] emitting [Result<Dog?>], representing either a valid dog, null if not found, or an error.
     */
    fun getDog(dogId: String): Flow<Result<Dog?>> {
        return if (!isValidUuid(dogId)) {
            // Immediately emit an error if the dogId is invalid
            flow {
                emit(Result.failure(Exception(errorMessages["INVALID_DOG_ID"])))
            }
        } else {
            dogRepository.getDog(dogId).map { repositoryResult ->
                if (repositoryResult.isFailure) {
                    // Propagate the repository error as-is
                    repositoryResult
                } else {
                    val fetchedDog = repositoryResult.getOrNull()
                    if (fetchedDog == null) {
                        // No dog found for the given ID
                        Result.success(null)
                    } else {
                        // Additional domain-level validations
                        if (!fetchedDog.active) {
                            return@map Result.failure(
                                Exception(errorMessages["INACTIVE_DOG"])
                            )
                        }
                        val age = fetchedDog.getAge()
                        if (age < 0 || age > 35) {
                            return@map Result.failure(
                                Exception(errorMessages["INVALID_AGE"])
                            )
                        }
                        if (!isMedicalInfoComplete(fetchedDog.medicalInfo)) {
                            return@map Result.failure(
                                Exception(errorMessages["INVALID_MEDICAL_INFO"])
                            )
                        }
                        // If all checks pass, return the dog
                        Result.success(fetchedDog)
                    }
                }
            }.catch { exception ->
                emit(Result.failure(exception))
            }
        }
    }

    /**
     * Retrieves a filtered and sorted list of a specific owner's dogs with pagination support.
     * Results are returned in a [Flow] wrapped with [Result]. This includes applying an optional
     * search query via the repository if present, filtering active dogs, sorting by name, and
     * finally paging the list.
     *
     * Steps:
     * 1. Validate ownerId format.
     * 2. Apply pagination parameters.
     * 3. Filter active dogs.
     * 4. Apply search query if provided (uses a hypothetical 'searchDogs' method in the repository).
     * 5. Sort dogs by name.
     * 6. Return Flow with paginated results.
     *
     * @param ownerId The unique ID of the owner whose dogs are being retrieved.
     * @param page The page index (0-based) for pagination.
     * @param pageSize The number of items to include per page.
     * @param searchQuery An optional search term to filter dogs by name or other criteria.
     * @return A [Flow] emitting [Result<List<Dog>>], handling data retrieval, sorting, and pagination.
     */
    fun getOwnerDogs(
        ownerId: String,
        page: Int,
        pageSize: Int,
        searchQuery: String?
    ): Flow<Result<List<Dog>>> {
        if (!isValidUuid(ownerId)) {
            return flow {
                emit(Result.failure(Exception(errorMessages["INVALID_OWNER_ID"])))
            }
        }
        if (page < 0 || pageSize <= 0) {
            return flow {
                emit(Result.failure(Exception(errorMessages["INVALID_PAGE_PARAMS"])))
            }
        }

        // If a search query is provided, assume a "searchDogs" method on the repository
        val baseFlow = if (!searchQuery.isNullOrBlank()) {
            dogRepository.searchDogs(ownerId, searchQuery)
        } else {
            dogRepository.getOwnerDogs(ownerId)
        }

        return baseFlow.map { repositoryResult ->
            if (repositoryResult.isFailure) {
                repositoryResult
            } else {
                // Retrieve the list from the success result
                val dogList = repositoryResult.getOrDefault(emptyList())
                // Filter only active dogs
                val activeDogs = dogList.filter { it.active }
                // Sort by name
                val sortedDogs = activeDogs.sortedBy { it.name }
                // Apply pagination
                val startIndex = page * pageSize
                val endIndex = (startIndex + pageSize).coerceAtMost(sortedDogs.size)
                if (startIndex >= sortedDogs.size) {
                    // No more items beyond this page
                    Result.success(emptyList())
                } else {
                    Result.success(sortedDogs.subList(startIndex, endIndex))
                }
            }
        }.catch { exception ->
            emit(Result.failure(exception))
        }
    }

    /**
     * Validates and saves dog information with comprehensive checks. Returns a suspended
     * [Result<Boolean>] indicating success or an error.
     *
     * Steps:
     * 1. Validate required fields completeness.
     * 2. Verify breed against valid breeds list.
     * 3. Validate medical information format.
     * 4. Check special instructions length and format.
     * 5. Verify age constraints.
     * 6. Save dog using the repository with error handling.
     * 7. Return Result with detailed status.
     *
     * @param dog The dog domain model to validate and save.
     * @return [Result<Boolean>] indicating operation success or failure with details.
     */
    suspend fun saveDog(dog: Dog): Result<Boolean> {
        return try {
            // 1. Basic field completeness checks have also been enforced in Dog's init block, but reaffirm here:
            if (dog.name.isBlank() || dog.id.isBlank() || dog.breed.isBlank()) {
                return Result.failure(Exception(errorMessages["SAVE_ERROR"]))
            }

            // 2. Verify breed
            if (!validBreeds.contains(dog.breed)) {
                return Result.failure(Exception(errorMessages["BREED_NOT_RECOGNIZED"]))
            }

            // 3. Validate medical info
            if (!isMedicalInfoComplete(dog.medicalInfo)) {
                return Result.failure(Exception(errorMessages["INVALID_MEDICAL_INFO"]))
            }

            // 4. Check special instructions (e.g. limit each to 200 chars)
            val tooLongInstruction = dog.specialInstructions.firstOrNull { it.length > 200 }
            if (tooLongInstruction != null) {
                return Result.failure(Exception(errorMessages["SPECIAL_INSTRUCTIONS_TOO_LONG"]))
            }

            // 5. Verify age constraints (dog should not have negative or unrealistic age)
            val age = dog.getAge()
            if (age < 0 || age > 35) {
                return Result.failure(Exception(errorMessages["INVALID_AGE"]))
            }

            // 6. Save dog in repository
            val saveResult = dogRepository.saveDog(dog)
            if (saveResult.isFailure) {
                return Result.failure(Exception(errorMessages["SAVE_ERROR"], saveResult.exceptionOrNull()))
            }
            saveResult
        } catch (ex: Exception) {
            Result.failure(Exception(errorMessages["SAVE_ERROR"], ex))
        }
    }

    /**
     * Handles dog profile deactivation with associated cleanup. This is typically done
     * when an owner or system admin decides to remove or temporarily disable a dog's
     * profile from active listings.
     *
     * Steps:
     * 1. Validate dogId format.
     * 2. Check current dog status.
     * 3. Record deactivation reason.
     * 4. Cancel associated future walks (placeholder logic).
     * 5. Execute soft delete in repository.
     * 6. Log audit trail.
     * 7. Return Result with operation status.
     *
     * @param dogId The unique string identifier of the dog to deactivate.
     * @param reason The reason provided for deactivation (used for auditing).
     * @return [Result<Boolean>] indicating whether the deactivation process succeeded.
     */
    suspend fun deactivateDog(dogId: String, reason: String): Result<Boolean> {
        return try {
            if (!isValidUuid(dogId)) {
                return Result.failure(Exception(errorMessages["INVALID_DOG_ID"]))
            }

            // 2. Check current dog status
            val dogFlow = dogRepository.getDog(dogId)
            val dogResult = dogFlow.first() // collect one emission
            if (dogResult.isFailure) {
                return dogResult.mapCatching { false }
            }
            val dogData = dogResult.getOrNull()
                ?: return Result.success(false) // if dog doesn't exist, return success(false)

            if (!dogData.active) {
                // If already inactive, we exit early with success
                return Result.success(true)
            }

            // 3. Record deactivation reason (Placeholder: log for now)
            // In a real system, this might be persisted in an audit trail or a separate table.
            // For demonstration:
            // e.g., logger.info("Deactivating dog $dogId due to: $reason")

            // 4. Cancel associated future walks (Placeholder)
            // Possibly we call a scheduling service or walk service to cancel upcoming sessions.
            // If that call fails, we can return an error:
            // return Result.failure(Exception(errorMessages["FUTURE_WALKS_CANCEL_ERROR"]))

            // 5. Execute soft delete in repository
            val deleteResult = dogRepository.deleteDog(dogId)
            if (deleteResult.isFailure) {
                return Result.failure(Exception(errorMessages["DEACTIVATION_ERROR"], deleteResult.exceptionOrNull()))
            }
            val deleted = deleteResult.getOrDefault(false)

            // 6. Log audit trail (Placeholder)
            // More robust systems might store an event in an auditing microservice or local log.
            // If this fails, we could return an error:
            // return Result.failure(Exception(errorMessages["AUDIT_TRAIL_ERROR"]))

            // 7. Return final operation status
            Result.success(deleted)
        } catch (ex: Exception) {
            Result.failure(Exception(errorMessages["DEACTIVATION_ERROR"], ex))
        }
    }

    /**
     * Checks whether a provided string conforms to a UUID format. This function can be
     * updated or swapped for library-based checks as needed.
     *
     * @param possibleUuid The string to validate.
     * @return True if the string is in a valid UUID format, false otherwise.
     */
    private fun isValidUuid(possibleUuid: String): Boolean {
        return runCatching {
            java.util.UUID.fromString(possibleUuid)
            true
        }.getOrDefault(false)
    }

    /**
     * Checks that a dog's medical information is sufficiently complete. This demonstration
     * simply checks if the map is not null. In real scenarios, you might enforce certain keys
     * or ensure that values are non-empty.
     *
     * @param medicalInfo The dog's medical info map to validate.
     * @return True if the info is (in this demonstration) non-empty or non-null.
     */
    private fun isMedicalInfoComplete(medicalInfo: Map<String, String>?): Boolean {
        return !medicalInfo.isNullOrEmpty()
    }
}